3rdparty/tools/rust/BUILD

```diff
@@ -0,0 +1,112 @@
+# Copyright 2023 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+# NB: TO get the links/hashes go to https://static.rust-lang.org/manifests.txt, then for the relevant
```

Comment:
Here's some alternative ideas that are likely much easier to maintain: 

- one `file`/`http_source` that's just the manifest `toml`, and then have a Python `adhoc_tool` or `shell_script` script (or 3 of them, to run in parallel) that does the downloads itself by reading that file?
- one `file`/`http_source` that's just `rustup`, convince it to install to specific directory (instead of the default `~/.rustup` one), and pull in the `rust-toolchain` file to have it select the version for us based on our repo config
  - I don't know if there's rustup binaries that can be easily installed directly, but there is dustup-init ones: https://rust-lang.github.io/rustup/installation/other.html
  - it looks like https://rust-lang.github.io/rustup/environment-variables.html lets us set `RUSTUP_HOME`, so we could have a `shell_command` that runs a command like the following, and has `output_directories=["cargo"]` as the current untarring command works:
    ```shell
    # get rustup to bootstrap the right version (I don't know if there's a better way)
    RUSTUP_HOME=rustup-home cargo version
    # we just need the single installed toolchain itself, not the rest of the rustup management
    mv rustup-home/toolchains/* cargo
    ```
  - Just a note on integrity/security, this would mean we're relying on the `http_source` hash of the `rustup-init`/`rustup` binary. It presumably does its own internal integrity checks on each artefact, based on the hashes in manifest toml(?). This seems fine to me! We're relying on `rustup` for security right now with the use of it and `rust-toolchain`.

Either of these seem like they'd be a dramatic quality-of-life improvement for bumping/maintaining Rust versions. If we can get the second working, that would be the smoothest (it's literally just the normal Rust flow), and would also start exploring a good feature for a real Rust backend: automatic rustc version management.

We could do them as follow up though, if we want this in sooner to unblock other work.

Reply:
I tried really hard to implement #2 but hit many brick walls.

Unfortunately I don't remember the specifics since that was a week or so ago 🫠

I can try again if I get time and report it here for posterity

Reply:
And for the record. I hate this one because calculating all those damn checksums and lengths was a pain

Reply:
Got it working!